### PR TITLE
Local tools

### PR DESCRIPTION
**Pull request by kaavee315**
Add local tools implementation

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -14,6 +14,7 @@ rich = ">=13.7.1,<14"
 importlib-metadata = ">=4.8.1,<5"
 cloup = ">=3.0.0"
 pyperclip = ">=1.8.2,<2"
+aenum = ">=3.1.15"
```
Comment (kaavee315) [inline Pipfile:17]: Need to remove

File: composio/client/__init__.py [hunk 2]
```diff
@@ -616,6 +618,18 @@ def get(  # type: ignore
         actions = actions or []
         apps = apps or []
         tags = tags or []
+        print("Apps - ", apps)
```
Comment (sohamganatra) [inline composio/client/__init__.py:621]: Should we be having print statements? Or do we don't have a logger in SDK?

File: composio/client/__init__.py [hunk 3]
```diff
@@ -616,6 +618,18 @@ def get(  # type: ignore
...
+        apps = [app for app in apps if not app.is_local]
+        actions = [action for action in actions if not action.is_local]
+        only_local_apps = len(apps) == 0 and len(actions) == 0 and (len(local_apps) > 0 or len(local_actions) > 0)
+        if only_local_apps:
+            local_items = self.local_handler.get_list_of_action_schemas(apps=local_apps, actions=local_actions, tags=tags)
+            return [self.model(**item) for item in local_items]
+
         if len(actions) > 0 and len(apps) > 0:
             raise ComposioClientError(
                 "Error retrieving Actions, Both actions and apps "
```
Comment (sohamganatra) [inline composio/client/__init__.py:635]: WTF does this error even mean?

**Comment by kaavee315** [comment]
**Shubhra** Add yourself to Pullflow

File: composio/local_tools/local_workspace/cmd_manager/actions/cmds.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    workspace_factory: WorkspaceManagerFactory = None
+    history_processor: HistoryProcessor = None
+
+    def set_workspace_and_history(self, workspace_factory: WorkspaceManagerFactory,
+                                  history_processor: HistoryProcessor):
+        self.workspace_factory = workspace_factory
+        self.history_processor = history_processor
+
+    def _setup(self, args: CreateFileRequest):
+        self.args = args
```
Comment (sohamganatra) [inline composio/local_tools/local_workspace/cmd_manager/actions/cmds.py:114]: Code repeat can be just captured sep.

**Review by sohamganatra — APPROVED** [review]

File: composio/local_tools/local_workspace/workspace/actions/create_workspace.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+from composio.local_tools.local_workspace.commons.local_docker_workspace import LocalDockerArgumentsModel
+from composio.local_tools.local_workspace.commons.local_docker_workspace import WorkspaceManagerFactory
+from composio.local_tools.local_workspace.commons.history_processor import HistoryProcessor
+
+
+logger = get_logger()
+
+
+class CreateWorkspaceRequest(BaseModel):
+    image_name: str = Field(..., description="the workspace is a docker container. To create a workspace"
```
[the remark this answers is not in the document]
Comment (shubhras01) [reply composio/local_tools/local_workspace/workspace/actions/create_workspace.py:14]: lets keep this here, we can add image-name in config and agent can take it from the config and give it to function

File: composio/local_tools/local_workspace/workspace/actions/create_workspace.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                                             "image name needs to be given, so container can start from the image")
+
+
+class CreateWorkspaceResponse(BaseModel):
+    workspace_id: str = Field(..., description="workspace-id for the created workspace")
+
+
+class CreateWorkspaceAction(Action):
+    """
+    creates a workspace, and returns workspace-id
```
[the remark this answers is not in the document]
Comment (shubhras01) [reply composio/local_tools/local_workspace/workspace/actions/create_workspace.py:24]: added

File: composio/local_tools/local_workspace/workspace/actions/create_workspace.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sohamganatra) [reply composio/local_tools/local_workspace/workspace/actions/create_workspace.py:14]: Yes, that's fine. But we can just add a default. For a new user, he has no idea what configs are possible what has to be given, for him to onboard quickly and start using this it helps to have a default value. You can always add a config value for your custom use case.

File: composio/client/__init__.py (hunk 3 above)
Comment (sohamganatra) [reply composio/client/__init__.py:635]: This is mainly gearer towards @kaavee315 
 CC: @shubhras01

Comment (kaavee315) [reply composio/client/__init__.py:635]: Did you miss reading the second line or still think the error is not clear? 
`Error retrieving Actions, Both actions and apps cannot be used as filters at the same time.`

File: composio/client/__init__.py (hunk 2 above)
Comment (shubhras01) [reply composio/client/__init__.py]: removed

**Pull request merged**